waspc/examples/todoApp/ext/Todo.js

```diff
@@ -98,12 +99,25 @@ const Tasks = (props) => {
 }
 
 const Task = (props) => {
+  const action = useAction(updateTaskIsDone, {
+    optimisticUpdates: [{
+      getQuery: () => [getTasks],
+      updateQuery: (updatedTask, oldTasks) => {
+        if (oldTasks === undefined) {
+          // cache is empty
+          return [updatedTask];
+        } else {
+          return oldTasks.map(task => task.id == updatedTask.id ? { ...task, ...updatedTask } : task)
+        }
+      }
```

Comment:
How should we behave when updating a query that does not yet exist in the cache?

Such a scenario can happen in the rare case when the users/developers wish to optimistically update something that is not in the current view. For example, in this case, the `oldData` will be `undefined`.

We can just document that the `updateQuery` should know how to handle the case when `oldTasks` is `undefined` (this is what I did here). Unfortunately, this approach does seem a bit janky (mainly because this behavior is something all queries could have in common).

Another option is ignoring such queries (i.e., not performing optimistic updates on queries not yet in the cache). This strategy would probably be OK in terms of UX (the query will get updated once the user lands on a view that features it, but it's less than ideal from the developer's perspective (a dev thinks they're updating something, and we're silently dropping it). We could mitigate the confusion by emphasizing that the `updateFn` _will not_ get called unless the provided query is already in the cache and optionally make this even more explicit by changing the function's name to something like `updateQueryIfCached`. That might be a bit extreme, though.

Reply:
Good idea. I'll go with it, seems like to most reasonable approach.

Reply:
Or, we could be explicit about two cases existing, by allowing them to define `updateCachedQuery` and/or `updateEmptyQuery`. So if they don't define the second one, we do default behaviour.

Reply:
But then they would often have to use the same function twice, right?

Reply:
Why the same function? I guess they would provide two different JS functions?

Reply:
In cases where the logic is the same (which could be common).

Reply:
Aha! Ok, well I think you have a better idea of how often that might be. Ok, whatever you think makes most sense!

Reply:
I'll leave it like this for now (let the programmers handle the case themselves). We can revise it at a later stage.